src/actions/sources.js

```diff
@@ -0,0 +1,64 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+// @flow
+import type { ThunkAction, SourceLoadingError } from 'firefox-profiler/types';
+import {
+  getDownloadRecipeForSourceFile,
+  parseFileNameFromSymbolication,
+} from '../utils/special-paths';
+import { assertExhaustiveCheck } from 'firefox-profiler/utils/flow';
+
+export function fetchSourceForFile(file: string): ThunkAction<Promise<void>> {
```

Comment:
For an optional follow-up:

TL;DR:
I think there's not much value in having this logic as a thunk action, and this could be handled in `BottomBox` completely, as it's called from there only.

More details:
IMO we're using thunk actions too much. They can be useful in some cases (especially being able to call `getState` can simplify a lot of things), but they also one big drawback: they don't "react" to a change of state, instead they need to be called directly. Also they force that we store things to the redux state even if that's not always necessary.

In this case this is exactly what we want to do: reacting to a change of the state `sourceView`.

My preference usually is to isolate the logic in a separate react component rendering `null` and reacting to state changes to trigger side effects.

In this specific case this could be done inside `BottomBox` directly instead of a separate react component because it's called from there only. But this could also be an unexported separate react component in the same file as `BottomBox`, as a child of the `BottomBox` component.

The separate react component could also be mounted as a child of ProfileViewer, and this could make the logic of the feature less complex.
Currently: double click -> open source view -> (sourceView state is changed + bottom box is rendered for this tab) -> BottomBox' didMount calls this thunk action which triggers the fetch.

Instead we could do:
double click -> open source view -> (sourceView state is changed + bottom box is rendered + SourceFetcher component fetches the file)

(Note: storing the current status to the redux state makes sense in this case.)

However for this case I believe this is mainly a matter of taste, I don't think one style or the other is better in terms of performance or even architecture for this case, because even if the current flow as outlined above is a bit complex, it's still well-defined and encapsulated with well-defined roles for each components. That's why I said this is optional.

Reply:
Interesting! I haven't fully understood all of what you said yet, I'll need to read it a few more times and think about it.

In general I definitely agree that we're using thunk actions too much, especially on the initialization path. And we're also often storing things in the state just so that it can get picked up by a selector, when it could just be a parameter and a utility function, for example.

Reply:
I've filed #3752 about this.